### PR TITLE
The problem of error occurring due to unsynchronized graphapi.

### LINKED ISSUE #64: Proxy error: Exception on /swap_path/eth/.../latest
Let me know if you need more info

```[2023-05-15 12:36:01,204] ERROR in app: Exception on /swap_path/eth/0x27a94869341838d5783368a8503fda5fbcd7987c/latest [GET]
Traceback (most recent call last):
  File "/home/rappie/.pyenv/versions/ityfuzz/lib/python3.10/site-packages/flask/app.py", line 2529, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/rappie/.pyenv/versions/ityfuzz/lib/python3.10/site-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/rappie/.pyenv/versions/ityfuzz/lib/python3.10/site-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/rappie/.pyenv/versions/ityfuzz/lib/python3.10/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/home/rappie/Desktop/ityfuzz.git/proxy/main.py", line 621, in swap_path
    return flask.jsonify(find_path_subgraph(network, token_address.lower(), block))
  File "/home/rappie/Desktop/ityfuzz.git/proxy/main.py", line 371, in find_path_subgraph
    hops = get_all_hops(token, network, block, known=set())
  File "/home/rappie/Desktop/ityfuzz.git/proxy/main.py", line 322, in get_all_hops
    hops[token] = get_pair(token, network, block)
  File "/home/rappie/.pyenv/versions/ityfuzz/lib/python3.10/site-packages/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/home/rappie/.pyenv/versions/ityfuzz/lib/python3.10/site-packages/retry/api.py", line 73, in retry_decorator
    return __retry_internal(partial(f, *args, **kwargs), exceptions, tries, delay, max_delay, backoff, jitter,
  File "/home/rappie/.pyenv/versions/ityfuzz/lib/python3.10/site-packages/retry/api.py", line 33, in __retry_internal
    return f()
  File "/home/rappie/Desktop/ityfuzz.git/proxy/main.py", line 266, in get_pair
    res = res["data"]
KeyError: 'data'
```

### PR DESCRIPTION
**Pull request by ByteSecurity**
Fix the issue of error reporting due to graphapi not synchronizing with the latest block.

1. Use the recommended last synchronized block by graphapi for requerying. 
2. in case of an error or inability to retrieve the trading pair, it will not crash and exit immediately.

fix #64

### CODE REVIEW

**Comment by ByteSecurity** [comment]
fix #100

**Comment by ByteSecurity** [comment]
fix #100 https://github.com/fuzzland/ityfuzz/issues/100#issuecomment-1623488188

**Comment by iVoider** [comment]
> fix #100 [#100 (comment)](https://github.com/fuzzland/ityfuzz/issues/100#issuecomment-1623488188)

Yes, there is no crash. But does the latest address I provided work for you? For me it just hangs on abi parsing part( maybe another issue). I tried with proxy and without it. Maybe python code should be also updated btw?

**Comment by iVoider** [comment]
Now, even contracts that had no "missing 'data'" problem are hanging.

**Comment by ByteSecurity** [comment]
`/ityfuzz/cli/target/release/cli -c BSC --onchain --onchain-block-number 29693964 -f -i --selfdestruct-oracle -p --run-forever --target 0x9e9b768174ef24233bf8ac2f4131f10ff5e72dee`

Regarding this parameter, I haven't encountered any suspension issues so far on my end. How long has this situation been occurring on your side?

**Comment by ByteSecurity** [comment]
`/ityfuzz/cli/target/release/cli -c BSC --onchain --onchain-block-number 29693964 -f -i --selfdestruct-oracle -p --run-forever --target 0x9e9b768174ef24233bf8ac2f4131f10ff5e72dee`

Regarding this parameter, I haven't encountered any suspension issues so far on my end. How long has this situation been occurring on your side?

**Comment by iVoider** [comment]
> /ityfuzz/cli/target/release/cli -c BSC --onchain --onchain-block-number 29693964 -f -i --selfdestruct-oracle -p --run-forever --target 0x9e9b768174ef24233bf8ac2f4131f10ff5e72dee

As long as I'm using "fix GetPairResponse json "null" error" commit, execution stucks on:

```
fetching code from a260e12d2b924cb899ae80bb58123ac3fee1e2f0
....
transferOwnership(address)
```
When revert to the first commit, it'll just panic as before.

**Comment by ByteSecurity** [comment]
> > /ityfuzz/cli/target/release/cli -c BSC --onchain --onchain-block-number 29693964 -f -i --selfdestruct-oracle -p --run-forever --target 0x9e9b768174ef24233bf8ac2f4131f10ff5e72dee
> 
> As long as I'm using "fix GetPairResponse json "null" error" commit, execution stucks on:
> 
> ```
> fetching code from a260e12d2b924cb899ae80bb58123ac3fee1e2f0
> ....
> transferOwnership(address)
> ```
> 
> When revert to the first commit, it'll just panic as before.

I found it too, wait for me

**Comment by iVoider** [comment]
> > > /ityfuzz/cli/target/release/cli -c BSC --onchain --onchain-block-number 29693964 -f -i --selfdestruct-oracle -p --run-forever --target 0x9e9b768174ef24233bf8ac2f4131f10ff5e72dee
> > 
> > 
> > As long as I'm using "fix GetPairResponse json "null" error" commit, execution stucks on:
> > ```
> > fetching code from a260e12d2b924cb899ae80bb58123ac3fee1e2f0
> > ....
> > transferOwnership(address)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > When revert to the first commit, it'll just panic as before.
> 
> I found it too, wait for me

It continued after hour of waiting. I'm still not sure if this ok, because other contracts that have pairs parsed correctly in main branch, also have this issue.

**Comment by ByteSecurity** [comment]
> It continued after hour of waiting. I'm still not sure if this ok, because other contracts that have pairs parsed correctly in main branch, also have this issue.
Another problem solved. Thank you, [iVoider](https://github.com/iVoider)

**Comment by ByteSecurity** [comment]
> 
> It continued after hour of waiting. I'm still not sure if this ok, because other contracts that have pairs parsed correctly in main branch, also have this issue.

Another problem solved. Thank you, [iVoider](https://github.com/iVoider).

File: src/evm/onchain/endpoints.rs [hunk 1]
```diff
@@ -715,7 +715,7 @@ impl OnChainConfig {
 impl OnChainConfig {
     fn get_pair(&self, token: &str, network: &str, block: &str) -> Vec<PairData> {
         let block_int;
-
+        println!("get pair {} {} {}", token, network, block);
```
Comment (Cloud10240) [inline src/evm/onchain/endpoints.rs:718]: Is this log necessary? Same question at line 760.

Comment (ByteSecurity) [reply src/evm/onchain/endpoints.rs:718]: When faced with a frozen interface or excessive refreshing,
 I chose to implement screen refreshing to let the user perceive that the program is actively performing tasks.

Comment (iVoider) [reply src/evm/onchain/endpoints.rs:718]: How about writing maxblock (latest synced by graph), also? So user could run consistently.

Comment (ByteSecurity) [reply src/evm/onchain/endpoints.rs:718]: @iVoider This is a temporary solution. In the future, we plan to develop our own graph server to synchronize the latest data for fuzzing purposes.

**Comment by shouc** [comment]
hey @ByteSecurity, please `git pull origin master` to use latest CI

**Review by Cloud10240 — APPROVED** [review]
LGTM

**Pull request merged**